doc/modules.md

```diff
@@ -0,0 +1,96 @@
+
+* Per-functionality targets
+* Improved build times for super projects
+
+## Implementation
+
+* All sub project targets are excluded from the default build, and
+  required targets are built only based on the dependencies.
+* Sub-libraries have a top-level header, e.g., lunchbox/log.h, and a
+  sub-directory with the implementation and CMakeLists, e.g.,
+  lunchbox/log/log.h and .cpp. The new common_add_library() cmake module
+  auto-generates the top-level header. New sub libraries set
+  NAME_OMIT_LIBRARY_HEADER.
+* The old common_library is deprecated, and stays during the transition
+  period.
+* Optional sub-targets are to be avoided and hidden by facades (esp. in
+  Lunchbox, see PersistentMap). If they are necessary, e.g, in hwsd,
+  downstream code needs to test for the target's availability (as done
+  e.g. in eq/server).
+* Servus will eventually be reintegrated as a lunchbox sub library.
+
+## Naming
+
+    common_add_library(Lunchbox::Log HEADERS logImpl.h PUBLIC_HEADERS log.h
+      SOURCES log.cpp DEPENDS PUBLIC Lunchbox::Bar PRIVATE Lunchbox::Foo)
+
+Generates: .../include/lunchbox/log.h, .../include/lunchbox/log/log.h,
+.../lib/libLunchboxLog.so
+
+    common_add_library(Lunchbox DEPENDS PUBLIC Lunchbox::Log)
+
+Generates: .../include/lunchbox.h
+
+## Examples
+
+    #include <zeroeq.h>
+    #include <lunchbox/log.h>
+
+### Desired Lunchbox Granularity
+
+    Lunchbox::Any
+    Lunchbox::Atomic
+    Lunchbox::Attic
+    Lunchbox::Bits
+    Lunchbox::Buffer
+    Lunchbox::Clock
+    Lunchbox::Condition
+    Lunchbox::DSO
+    Lunchbox::File
+    Lunchbox::Future
+    Lunchbox::Launcher
+    Lunchbox::Lock (lock, spinlock, scoped, monitor, queue, lockable)
+    Lunchbox::LockFree
+    Lunchbox::Log
+    Lunchbox::MPI
+    Lunchbox::MemoryMap
+    Lunchbox::Monitor
+    Lunchbox::OMP
+    Lunchbox::OS
+    Lunchbox::PersistentMap
+    Lunchbox::Plugin
+    Lunchbox::Process (daemon+launcher)
+    Lunchbox::RNG
+    Lunchbox::Servus
+    Lunchbox::TLS
+    Lunchbox::Test
+    Lunchbox::Thread
+    Lunchbox::URI
+    Lunchbox::OrderedIntervalSet (was UnorderedIntervalSet)
+
+## Issues
+
+### 1: How do handle optional targets?
+
+_Resolution: Avoid them as possible, downstream projects need to check
+with if(TARGET) in cmake:_
+
+We could (re)use the COMPONENTS parameter in common_find_package(), but
+this would overload semantics and add magic. Right now we only have a
+few use cases (hwsd, ZeroEQ Qt support). We might decide to improve
+common_find_package() later if needed.
+
+Alternatively we could generate a package config for each sub library,
+so that downstream projects check for each sub library individually,
+e.g., common_find_package(Lunchbox::Log REQUIRED). In consequence this
+means that each sub library becomes a separate git project: find_package
+is designed to find a project, and we eventually want to extend the
+common_find_package syntax to include a git url and SHA to get rid of
+the redundant .gitsubprojects.
```

Comment:
I haven't given much thought about how the merging of common_find_package and .gitsubproject will look like in practice. Maybe it should be a new command, that calls  common_find_package internally to avoid the repetition but does not cause confusion with finding system packages. In any case, I don't think it's a blocker, and that it implies creating separate git projects. I does point to something interesting though, that maybe each library could be versionned separately.
I agree with Juan that what a package is is ill-defined in general. For instance, boost or qt come as several pacakges on linux, but as one install on macports. Also Poppler, which can be considered as one small project, but is split into Poppler, poppler-qt, poppler-glib, etc...
COMPONENTS finding look like the right thing to do at first but they are a pain to handle in practice. I know it from writing a FindPoppler.cmake with separate components and from the problems that we have (had) with finding all those boost components. The way Qt works now is much better IMO.
On the contrary, there are plenty of such optional targets: ServusQt, DeflectServer (today part of Deflect, but should be separated), probably 90% of the upcoming Lunchbox targets, etc.